include/tvm/function/BasicLinearFunction.h

```diff
@@ -93,7 +101,24 @@ template<typename CstDerived, typename... Derived>
 BasicLinearFunction::BasicLinearFunction(const utils::AffineExpr<CstDerived, Derived...> & aff)
 : LinearFunction(static_cast<int>(std::get<0>(aff.linear()).matrix().rows()))
 {
-  add(aff.linear(), std::make_index_sequence<sizeof...(Derived)>{});
+  using Indices = std::make_index_sequence<sizeof...(Derived)>;
+  tvm::internal::VariableCountingVector v;
+  internal::addVar(v, aff.linear(), Indices{});
+  const auto & vars = v.variables();
+  const auto & simple = v.simple();
+  for(int i = 0; i < vars.numberOfVariables(); ++i)
+  {
+    if(!simple[i])
+    {
+      addVariable(vars[i], true);
+    }
+  }
+  for(auto & J : jacobian_)
+  {
+    J.second.setZero();
+    J.second.properties({tvm::internal::MatrixProperties::Constness(true)});
+  }
```

Comment:
`simple[i]` basically means that `vars[i]` is the result of the first `add` on `v` for this variable.

If a variable is not simple, meaning that it is the aggregation of several subvariables that made it grow, it is difficult to track how the columns of the corresponding jacobian will be initialized. For a not simple variable, we therefore add the variable manually (l. 113) and initialize it to zero (l. 118). It will then be filled by using += (l. 147).

A simple variable is added and its jacobian initialized directly in `add` (l. 153-154). This avoids the initial initialization to 0 and should be the usual path.

Reply:
I added the same kind of explanation in 2c92d1bc280a7e8c8726f8d713b6a9c212e540fc